File: 16. hot100/[46]全排列.java
//给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。 
//
// 
//
// 示例 1： 
//
// 
//输入：nums = [1,2,3]
//输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
// 
//
// 示例 2： 
//
// 
//输入：nums = [0,1]
//输出：[[0,1],[1,0]]
// 
//
// 示例 3： 
//
// 
//输入：nums = [1]
//输出：[[1]]
// 
//
// 
//
// 提示： 
//
// 
// 1 <= nums.length <= 6 
// -10 <= nums[i] <= 10 
// nums 中的所有整数 互不相同 
// 
// Related Topics 回溯算法 
// 👍 1339 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    private void swap(int[] box, int i, int j) {
        int tmp = box[i];
        box[i] = box[j];
        box[j] = tmp;
    }
    private void backTrace(int[] box, int i, List<List<Integer>> ans) {
        final int N = box==null ? 0:box.length;
        if (i == N) {
            ans.add(new ArrayList<>());
            for (int x : box) {
                ans.get(ans.size()-1).add(x);
            }
        }
        if (i >= N) {
            return;
        }
        for (int j = i; j < N; j++) {
            swap(box, i, j);
            backTrace(box, i+1,  ans);
            swap(box, i, j);
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        final int N = nums==null ? 0:nums.length;

        List<List<Integer>> ans = new ArrayList<>();
        backTrace(nums, 0,  ans);
        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 11. backTrace/[46]全排列.py
# 给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。 
# 
#  
# 
#  示例 1： 
# 
#  
# 输入：nums = [1,2,3]
# 输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
#  
# 
#  示例 2： 
# 
#  
# 输入：nums = [0,1]
# 输出：[[0,1],[1,0]]
#  
# 
#  示例 3： 
# 
#  
# 输入：nums = [1]
# 输出：[[1]]
#  
# 
#  
# 
#  提示： 
# 
#  
#  1 <= nums.length <= 6 
#  -10 <= nums[i] <= 10 
#  nums 中的所有整数 互不相同 
#  
# 
#  Related Topics 数组 回溯 👍 2743 👎 0

from typing import List
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def backtrace(self, nums, i,used,  box, ans):
        N = 0 if not nums else len(nums)
        if len(box)==N:
            ans.append([])
            for x in box:
                ans[-1].append(x)

        if i>=N:
            return

        for j in range(N):
            if not used[j]:
                used[j] = True
                box.append(nums[j])
                self.backtrace(nums, j, used, box, ans)
                box.pop()
                used[j] =False

    def permute(self, nums: List[int]) -> List[List[int]]:
        box = []
        ans = []
        used = [False]*len(nums)
        self.backtrace(nums, 0,used, box, ans)
        return ans

# leetcode submit region end(Prohibit modification and deletion)
